### ISSUE TITLE
DataSource.countColumns returns invalid number of columns for nested objects.

### ISSUE THREAD

**Issue by lewisjb** [handsontable/handsontable#3958]
### Description
Line 194 in src/dataSource:
https://github.com/handsontable/handsontable/blob/master/src/dataSource.js#L194
(`result = Object.keys(this.data[0]).length;`)

Won't/doesn't work for https://docs.handsontable.com/0.29.1/tutorial-data-sources.html#example5

```javascript
nestedObjects = [
  {id: 1, name: {first: "Ted", last: "Right"}, address: ""},
  {id: 2, address: ""}, // HOT will create missing properties on demand
  {id: 3, name: {first: "Joan", last: "Well"}, address: ""}
]
```

`Object.keys` on that will return `['id', 'name', 'address']`.

Using `DataSource.hot.countCols` on the other hand, works. So copying the logic / calling that function would solve the issue

**Comment by AMBudnik** [comment]
related https://github.com/handsontable/handsontable/issues/3959

**Comment by swistach** [comment]
Hi @lewisjb 
It's not a bug.
Difference between column count in dataMap and dataSource is construction of data in each mechanism.
dataMap (hot.countCols) contains data always as an array with X and Y axis, and we need to show the nested object as an array, which should be:
[id, name.first, name.last, address]

But dataSource contains object, and at first level has only 3 keys:
{id, name{}, address}.

For proof that it works properly, you can run hot.getSourceData and check what was returned. If returned data has the same structure as during initialization then everything should be ok.

If something isn't clear, send us your questions. We'll try to help :)

**Comment by AMBudnik** [comment]
Thanks for clarifying @swistach 
I guess we can close this issue.

**Comment by lewisjb** [comment]
@swistach Apologies for the late response.

I am a little confused as to how this is the intended behaviour, I get that dataSource contains an object with only 3 keys at the first level, but these keys are in no way 'columns' (as the name of the function suggests). Is it just that the function was poorly named? Or am I mis-understanding something?

@AMBudnik was right in that this is related to #3959, if this is the intended behaviour for some reason, then I believe the real bug is that the wrong function is used here: https://github.com/handsontable/handsontable/blob/master/src/dataSource.js#L48

In retrospect I went about this the wrong way, instead of creating an issue about the actual problem; I created issues about what I found to be the causes.

The real issue is that when you delete a row that uses nested objects, it can't be undone properly.
Example: http://jsfiddle.net/59ov3fdv/1/
If you right-click->Remove Row, then undo, only the first column comes back. After making the proposed adjustments in this and the related issue, the problem was resolved.

**Comment by AMBudnik** [comment]
@swistach, Lewis is right here. In his demo when we delete the 3rd row and undo the change our data is lost.

![issue](https://cloud.githubusercontent.com/assets/10757813/21809569/9e797146-d748-11e6-9135-37442ed322b3.gif)

I guess it is only a PRO bug but we can have it here opened. What do you think?

**Comment by wojciechczerniak** [comment]
@AMBudnik Fixed in v8? https://jsfiddle.net/6fa8dj07/

![Jul-07-2020 10-40-15](https://user-images.githubusercontent.com/4718998/86750506-4bc61700-c03e-11ea-944d-46b190307f44.gif)

**Comment by AMBudnik** [comment]
true @wojciechczerniak when we use 'undo' data's back.

It was broken in `7.4.2` https://jsfiddle.net/3e1L8qvr/ and `8.0.0-beta1` https://jsfiddle.net/h2L1u8n7/ so we can call it solved by `beta2`.

**Comment by pnowak** [comment]
Hi @lewisjb 

I have good news! We've just released 8.0.0-beta2 that fixes the mentioned issue. We won't be closing this ticket now. It will be closed after we release the final 8.0.0 version of Handsontable.

Here’s a full list of changes https://github.com/handsontable/handsontable/releases/tag/8.0.0-beta.2
NPM https://www.npmjs.com/package/handsontable/v/8.0.0-beta.2
 
All feedback's appreciated. Thank you for your input.

**Comment by pnowak** [comment]
The issue is still fixed in 8.0.0-rev1 - https://jsfiddle.net/uksf64nc/

**Comment by AMBudnik** [comment]
Hi @lewisjb 

I’m glad to announce that Handsontable v8.0.0 that solves this issue is live!   I highly encourage you to test the new version and leave some feedback. I hope that the amount of work that we put on the v8 will help you to improve your project.

CDN https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.css
https://cdn.jsdelivr.net/gh/handsontable/handsontable@latest/dist/handsontable.full.js
NPM npm i handsontable
Release Notes https://github.com/handsontable/handsontable/releases
